{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Phylotrackpy with synchronous generations example (completed)\n",
    "\n",
    "This notebook contains a simple evolving system with synchronous generations, which is typical for most evolutionary computing systems. \n",
    "That is, each generation, a population of candidate solutions are evaluated and selected to reproduce. \n",
    "This process repeats for a desired number of generations. \n",
    "\n",
    "In this example, we implement the one-max problem where individuals are binary strings, and a solution has a genome of all 1s. \n",
    "The example code in this notebook incorporates phylogeny tracking.\n",
    "Below, we link to some existing examples and the `phylotrackpy` documentation to \n",
    "for your reference. \n",
    "If you're working on this during the tutorial at ALife 2024, feel free to ask for our help in understanding the code! \n",
    "We're happy to walk you through anything!  \n",
    "\n",
    "## Helpful reference material\n",
    "\n",
    "- `phylotrackpy` documentation: <https://phylotrackpy.readthedocs.io/en/latest/introduction.html>\n",
    "- `phylotrackpy` GitHub repository: <https://github.com/emilydolson/phylotrackpy>\n",
    "- Example code from an ALife 2023 tutorial: <https://github.com/emilydolson/alife-phylogeny-tutorial/blob/main/perfect_tracking_final.ipynb> \n",
    "  - Scroll down to the `phylotrackpy` heading!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup\n",
    "\n",
    "First, install required Python packages for this example (e.g., phylotrackpy).\n",
    "If you are running this locally, we recommend making a Python virtual environment to keep your local Python installation clean: \n",
    "\n",
    "```\n",
    "python -m venv venv\n",
    "source venv/bin/activate\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!python -m pip install -r requirements.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "from phylotrackpy import systematics\n",
    "import random\n",
    "import polars as pl\n",
    "\n",
    "# Seed random number generator\n",
    "random.seed(8)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## System implementation\n",
    "\n",
    "Here, the `Organism` class (below) defines candidate solutions with binary genomes, an evaluated fitness value, and a taxon ID. \n",
    "The taxon ID member variable keeps track of the organism's taxon in the phylogeny, which is helpful for `phylotrackpy`'s phylogeny tracker.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Organism:\n",
    "    def __init__(\n",
    "        self,\n",
    "        num_genes:int = 10,\n",
    "        randomize_genome:bool = False\n",
    "    ):\n",
    "        # Genomes are vectors of binary values\n",
    "        self.genome = [bool(random.randint(0, 1)) if randomize_genome else False for _ in range(num_genes)]\n",
    "        # Evaluate initial fitness\n",
    "        self.EvalFitness()\n",
    "        # Organisms keep track of their taxon id (useful for phylogeny tracking)\n",
    "        self.taxon_id = None\n",
    "\n",
    "    @classmethod\n",
    "    def FromGenome(cls, genome:list):\n",
    "        # Create new organism from a given genome.\n",
    "        org = cls(\n",
    "            num_genes = 0,\n",
    "            randomize_genome = False\n",
    "        )\n",
    "        org.SetGenome(genome)\n",
    "        org.EvalFitness()\n",
    "        return org\n",
    "\n",
    "    def GetGenome(self):\n",
    "        return self.genome\n",
    "\n",
    "    def SetGenome(self, genome:list):\n",
    "        self.genome = [gene for gene in genome]\n",
    "        # Evaluate fitness after updating genome\n",
    "        self.EvalFitness()\n",
    "\n",
    "    def GetFitness(self):\n",
    "        return self.fitness\n",
    "\n",
    "    def EvalFitness(self):\n",
    "        self.fitness = sum(self.genome)\n",
    "        return self.fitness\n",
    "\n",
    "    def GetTaxonID(self):\n",
    "        return self.taxon_id\n",
    "\n",
    "    def SetTaxonID(self, id):\n",
    "        self.taxon_id = id\n",
    "\n",
    "    def Mutate(self, per_site_mut_rate:float=0.01):\n",
    "        num_muts = 0\n",
    "        for gene_i in range(len(self.genome)):\n",
    "            if (random.random() <= per_site_mut_rate):\n",
    "                self.genome[gene_i] = not self.genome[gene_i]\n",
    "                num_muts += 1\n",
    "        # Update fitness evaluation after mutations\n",
    "        self.EvalFitness()\n",
    "        return num_muts\n",
    "\n",
    "\n",
    "def TournamentSelect(num_parents:int, population:list, tourn_size:int = 4):\n",
    "    parent_ids = [None for _ in range(num_parents)]\n",
    "    candidate_ids = [i for i in range(len(population))]\n",
    "    for i in range(num_parents):\n",
    "        tourn_participants = random.sample(candidate_ids, tourn_size)\n",
    "        parent_ids[i] = max([(population[idx].GetFitness(), idx) for idx in tourn_participants])[1]\n",
    "    return parent_ids\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here, we implement a basic evolutionary computing loop solving the one-max problem to demonstrate how `phylotrackpy` can be integrated into a traditional evolutionary computing context."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---Update 0---\n",
      "  Max fitness=0\n",
      "---Update 100---\n",
      "  Max fitness=87\n",
      "---Update 200---\n",
      "  Max fitness=89\n",
      "---Update 300---\n",
      "  Max fitness=88\n",
      "---Update 400---\n",
      "  Max fitness=89\n",
      "---Update 500---\n",
      "  Max fitness=90\n",
      "---Update 600---\n",
      "  Max fitness=88\n",
      "---Update 700---\n",
      "  Max fitness=90\n",
      "---Update 800---\n",
      "  Max fitness=89\n",
      "---Update 900---\n",
      "  Max fitness=88\n",
      "---Update 1000---\n",
      "  Max fitness=90\n"
     ]
    }
   ],
   "source": [
    "pop_size = 500\n",
    "generations = 1000\n",
    "per_site_mut_rate = 0.05\n",
    "genome_length = 100\n",
    "tournament_size = 8\n",
    "print_resolution = 100\n",
    "summary_output_resolution = 10\n",
    "phylo_snapshot_resolution = 1000\n",
    "\n",
    "assert(pop_size > 0)\n",
    "assert(generations > 0)\n",
    "assert(per_site_mut_rate >= 0 and per_site_mut_rate <= 1.0)\n",
    "assert(genome_length > 0)\n",
    "assert(print_resolution > 0)\n",
    "assert(summary_output_resolution > 0)\n",
    "\n",
    "# Create a new Systematics object to track the population's phylogeny\n",
    "# - Taxa in the phylogeny will be defined by genotypes (i.e., offspring with the same genotype as their parent will belong to the same taxon).\n",
    "phylo_tracker = systematics.Systematics(\n",
    "    lambda org: org.GetGenome(),\n",
    "    store_pos = False\n",
    ")\n",
    "# Configure snapshots to output basic information about a taxon\n",
    "phylo_tracker.add_snapshot_fun(\n",
    "    lambda tax: str(list(map(int,tax.get_info()))).replace(\",\", \" \"),\n",
    "    \"genome\",\n",
    "    \"Taxon's genome\"\n",
    ")\n",
    "phylo_tracker.add_snapshot_fun(\n",
    "    lambda tax: str(sum(tax.get_info())),\n",
    "    \"fitness\",\n",
    "    \"Number of ones in taxon genome\"\n",
    ")\n",
    "# Initialize phylogeny tracker's update to 0\n",
    "phylo_tracker.set_update(0)\n",
    "\n",
    "# Create a list to hold phylo metrics over time\n",
    "data = []\n",
    "\n",
    "# Create initial population\n",
    "population = [Organism(num_genes=genome_length, randomize_genome=False) for _ in range(pop_size)]\n",
    "# Add initial population to phylogeny tracker\n",
    "for indiv in population:\n",
    "    indiv.SetTaxonID(phylo_tracker.add_org(indiv))\n",
    "\n",
    "for gen in range(0, generations+1):\n",
    "    if (gen % print_resolution) == 0:\n",
    "        print(f\"---Update {gen}---\")\n",
    "        print(f\"  Max fitness={max([indiv.GetFitness() for indiv in population])}\")\n",
    "\n",
    "    # Set phylo tracker's update to current generation\n",
    "    phylo_tracker.set_update(gen)\n",
    "\n",
    "    # Individuals are evaluated immediately after reproduction, so no need to evaluate\n",
    "    # the full population here. In other systems, population evaulation would likely happen here.\n",
    "\n",
    "    # Select parents to reproduce.\n",
    "    parent_ids = TournamentSelect(\n",
    "        num_parents = pop_size,\n",
    "        population = population,\n",
    "        tourn_size = tournament_size\n",
    "    )\n",
    "\n",
    "    # Create next generation's population by copying parents, mutating offspring,\n",
    "    # and adding offspring to the phylogeny.\n",
    "    next_pop = []\n",
    "    for parent_id in parent_ids:\n",
    "        next_pop.append(Organism.FromGenome(population[parent_id].GetGenome()))\n",
    "        next_pop[-1].Mutate(per_site_mut_rate)\n",
    "        offspring_taxon_id = phylo_tracker.add_org(next_pop[-1], population[parent_id].GetTaxonID())\n",
    "        next_pop[-1].SetTaxonID(offspring_taxon_id)\n",
    "\n",
    "    # Snapshot phylogeny\n",
    "    if (gen % phylo_snapshot_resolution) == 0:\n",
    "        phylo_tracker.snapshot(f\"phylo_snapshot_{gen}.csv\")\n",
    "\n",
    "    # Record data for this update\n",
    "    if (gen % summary_output_resolution) == 0:\n",
    "        data.append({\n",
    "            \"generation\": gen,\n",
    "            \"max_fit\": max([indiv.GetFitness() for indiv in population]),\n",
    "            \"phylo_num_taxa\": phylo_tracker.get_num_taxa(),\n",
    "            \"phylo_sum_pairwise_dist\": phylo_tracker.get_sum_pairwise_distance(False),\n",
    "            \"phylo_variance_pairwise_dist\": phylo_tracker.get_variance_pairwise_distance(False),\n",
    "            \"phylo_phylogenetic_diversity\": phylo_tracker.get_phylogenetic_diversity(),\n",
    "            \"phylo_mrca_depth\": phylo_tracker.mrca_depth()\n",
    "        })\n",
    "\n",
    "    # Remove parents from phylogeny\n",
    "    for indiv in population:\n",
    "        phylo_tracker.remove_org(indiv.GetTaxonID())\n",
    "    population = next_pop\n",
    "\n",
    "phylo_tracker.snapshot(\"phylo_snapshot_final.csv\")\n",
    "summary_df = pl.DataFrame(data)\n",
    "summary_df.write_csv(\"phylo_summary.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pyenv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
