{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Phylotrackpy with asynchronous generations example\n",
    "\n",
    "This notebook contains a very simple evolving system with asynchronous generations \n",
    "wherein organisms reproduce as they accumulate the requisite resources.\n",
    "Each organism's genome is a binary string, and organism's accumulate resources as \n",
    "a function of the proportion of 1s in their genome. \n",
    "I.e., this is a one-max environment!  \n",
    "\n",
    "The code in this notebook does not incorporate phylogeny tracking (see [this notebook](https://github.com/amlalejini/alife-2024-phylo-tutorial/blob/main/notebooks/phylotrackpy-async-gen-example-completed.ipynb) for \n",
    "a version of this code that _does_ already incorporate phylogeny tracking).\n",
    "Think of this notebook as an opportunity to play around with how you could integrate\n",
    "phylogeny tracking with `phylotrackpy` into an existing system.\n",
    "Throughout the code, we have left comments where phylogeny tracking code should be added.\n",
    "Below, we link to some existing examples and the `phylotrackpy` documentation to \n",
    "for your reference to get tracking working. \n",
    "If you're working on this during the tutorial at ALife 2024, feel free to ask for our help! \n",
    "We're happy to walk you through anything!  \n",
    "\n",
    "## Helpful reference material\n",
    "\n",
    "- `phylotrackpy` documentation: <https://phylotrackpy.readthedocs.io/en/latest/introduction.html>\n",
    "- `phylotrackpy` GitHub repository: <https://github.com/emilydolson/phylotrackpy>\n",
    "- Example code from an ALife 2023 tutorial: <https://github.com/emilydolson/alife-phylogeny-tutorial/blob/main/perfect_tracking_final.ipynb> \n",
    "  - Scroll down to the `phylotrackpy` heading!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, install required Python packages for this example. (e.g., `phylotrackpy`). \n",
    "If you are running this locally, we recommend making a python virtual environment \n",
    "to keep your local python installation clean:\n",
    "\n",
    "```\n",
    "python -m venv venv\n",
    "source venv/bin/activate\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!python3 -m pip install -r requirements.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from phylotrackpy import systematics\n",
    "import random\n",
    "import polars as pl\n",
    "\n",
    "# Seed random number generator\n",
    "random.seed(8)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## System implementation\n",
    "\n",
    "The `Organism` class (below) defines organisms with a binary genome, resource count, and taxon id.\n",
    "The taxon id member keeps track of the organism's taxon in the phylogeny (which is helpful for `phylotrackpy`'s phylogeny tracker)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Organism:\n",
    "    def __init__(\n",
    "        self,\n",
    "        num_genes:int = 10,\n",
    "        randomize_genome:bool = False,\n",
    "        init_resources:float = 0.0\n",
    "    ):\n",
    "        # Genomes are vectors of binary values\n",
    "        self.genome = [bool(random.randint(0, 1)) if randomize_genome else False for _ in range(num_genes)]\n",
    "        # Organisms have resources that determine whether they can reproduce\n",
    "        self.resources = init_resources\n",
    "        # Organisms keep track of their taxon id (useful for phylogeny tracking)\n",
    "        self.taxon_id = None\n",
    "\n",
    "    @classmethod\n",
    "    def FromGenome(cls, genome:list):\n",
    "        # Create new organism from a given genome.\n",
    "        org = cls(\n",
    "            num_genes = 0,\n",
    "            randomize_genome = False,\n",
    "            init_resources = 0.0\n",
    "        )\n",
    "        org.SetGenome(genome)\n",
    "        return org\n",
    "\n",
    "    def GetGenome(self):\n",
    "        return self.genome\n",
    "\n",
    "    def SetGenome(self, genome:list):\n",
    "        self.genome = [gene for gene in genome]\n",
    "\n",
    "    def GetResources(self):\n",
    "        return self.resources\n",
    "\n",
    "    def SetResources(self, amount:float):\n",
    "        self.resources = amount\n",
    "\n",
    "    def IncResources(self, amount:float=1.0):\n",
    "        self.resources += amount\n",
    "\n",
    "    def DecResources(self, amount:float=1.0):\n",
    "        self.resources -= amount\n",
    "\n",
    "    def GetTaxonID(self):\n",
    "        return self.taxon_id\n",
    "\n",
    "    def SetTaxonID(self, id):\n",
    "        self.taxon_id = id\n",
    "\n",
    "    def Mutate(self, per_site_mut_rate:float=0.01):\n",
    "        num_muts = 0\n",
    "        for gene_i in range(len(self.genome)):\n",
    "            if (random.random() <= per_site_mut_rate):\n",
    "                self.genome[gene_i] = not self.genome[gene_i]\n",
    "                num_muts += 1\n",
    "        return num_muts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The \"world\" implementation is below. Organisms exist in a \"well-mixed\" environment (the `world` list).\n",
    "On each update, an organism is chosen at random to \"execute\", gaining resources proportional to the number of 1s in its genome. \n",
    "That organism then reproduces if it has sufficient resources. \n",
    "\n",
    "**We've marked locations where phylogeny tracking code needs to be added with \"TUTORIAL TASK\" comments!**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# World parameters\n",
    "max_world_size = 500           # Determines maximum number of organisms in world\n",
    "updates = 20000                 # How long to run system?\n",
    "repro_cost = 10.0              # Cost of reproduction\n",
    "max_resource_intake = 5        # Maximum amount of resources an organism can gain when \"executed\"\n",
    "per_site_mut_rate = 0.05       # Per-site mutation rate for offspring\n",
    "genome_length = 100            # Length of organism genomes\n",
    "print_resolution = 100         # Print world status at this interval\n",
    "summary_output_resolution = 10 # Interval to save summary information\n",
    "\n",
    "assert(max_world_size > 0)\n",
    "assert(updates > 0)\n",
    "assert(per_site_mut_rate >= 0 and per_site_mut_rate <= 1.0)\n",
    "assert(genome_length > 0)\n",
    "\n",
    "'''\n",
    "TUTORIAL TASK: Initialization\n",
    "- (1) Create a new systematics object here to track the population's phylogeny\n",
    "- (2) Add snapshot functions that capture basic information about a taxon (ask for clarification if you're not sure what this means)\n",
    "- (3) Initialize the systematics object's update to 0\n",
    "'''\n",
    "\n",
    "# Create a list to hold phylo metrics over time\n",
    "data = []\n",
    "\n",
    "# Create population seeded with initial organism\n",
    "world = [Organism(num_genes=genome_length, randomize_genome=False)]\n",
    "\n",
    "'''\n",
    "TUTORIAL TASK: Add initial organism to phylogeny\n",
    "- After adding the initial organism to the phylogeny, store its taxon id on the organism\n",
    "'''\n",
    "\n",
    "# Execute world for configured number of updates\n",
    "for update in range(0, updates+1):\n",
    "    if (update % print_resolution) == 0:\n",
    "        print(f\"---Update {update}---\")\n",
    "        print(f\"  Population size={len(world)}\")\n",
    "        print(f\"  Max fitness={max([sum(world[i].GetGenome()) for i in range(len(world))])}\")\n",
    "\n",
    "    '''\n",
    "    TUTORIAL TASK: Update the systematics object's update to current update\n",
    "    '''\n",
    "\n",
    "    # Get ID of individual to \"execute\"\n",
    "    pop_id = random.randrange(0, len(world)) if len(world) > 1 else 0\n",
    "    cur_org = world[pop_id]\n",
    "\n",
    "    # Calculate resource gain\n",
    "    fit = sum(cur_org.GetGenome())\n",
    "    resource_gain = max((fit / genome_length) * max_resource_intake, 0.1)\n",
    "    cur_org.IncResources(resource_gain)\n",
    "\n",
    "    # Check if organism has resources to reproduce\n",
    "    if cur_org.GetResources() >= repro_cost:\n",
    "        # Current organism pays cost of reproduction\n",
    "        cur_org.DecResources(repro_cost)\n",
    "        # Create offspring as copy of current organism\n",
    "        offspring = Organism.FromGenome(cur_org.GetGenome())\n",
    "        mut_count = offspring.Mutate(per_site_mut_rate)\n",
    "\n",
    "        '''\n",
    "        TUTORIAL TASK: Add new offspring to phylogeny\n",
    "        - (1) Add new organism to phylogeny\n",
    "        - (2) Update offspring's taxon id\n",
    "        '''\n",
    "\n",
    "        # Place offspring into world\n",
    "        if len(world) < max_world_size:\n",
    "            # If world not full, just append.\n",
    "            world.append(offspring)\n",
    "        else:\n",
    "            # Otherwise, random replacement.\n",
    "            offspring_loc = random.randint(0, len(world)-1)\n",
    "\n",
    "            '''\n",
    "            TUTORIAL TASK: Remove organism that has been chosen from replacement from the phylogeny\n",
    "            '''\n",
    "\n",
    "            # Offspring replaces organism at chosen location.\n",
    "            world[offspring_loc] = offspring\n",
    "\n",
    "    # Record data for this update\n",
    "    if (update % summary_output_resolution) == 0:\n",
    "        '''\n",
    "        TUTORIAL TASK: Add some phylogeny metrics to the data being saved\n",
    "        '''\n",
    "        data.append({\n",
    "            \"update\": update,\n",
    "            \"pop_size\": len(world)\n",
    "        })\n",
    "\n",
    "\n",
    "'''\n",
    "TUTORIAL TASK: Output final phylogeny snapshot\n",
    "'''\n",
    "summary_df = pl.DataFrame(data)\n",
    "summary_df.write_csv(\"summary.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data visualization\n",
    "\n",
    "Once we have phylogenies, we can visualize them! \n",
    "\n",
    "TODO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
