{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a target=\"_blank\" href=\"https://colab.research.google.com/github/amlalejini/alife-2024-phylo-tutorial/blob/main/notebooks/phylotrackpy-sync-gen-example.ipynb\">\n",
    "  <img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/>\n",
    "</a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Phylotrackpy with synchronous generations example\n",
    "\n",
    "This notebook contains a simple evolving system with synchronous generations, which is typical for most evolutionary computing systems. \n",
    "That is, each generation, a population of candidate solutions are evaluated and selected to reproduce. \n",
    "This process repeats for a desired number of generations. \n",
    "\n",
    "In this example, we implement the one-max problem where individuals are binary strings, and a solution has a genome of all 1s. \n",
    "\n",
    "The code in this notebook does not incorporate phylogeny tracking (see [this notebook](https://github.com/amlalejini/alife-2024-phylo-tutorial/blob/main/notebooks/phylotrackpy-sync-gen-example-completed.ipynb) for \n",
    "a version of this code that _does_ already incorporate phylogeny tracking).\n",
    "Think of this notebook as an opportunity to play around with how you could integrate\n",
    "phylogeny tracking with `phylotrackpy` into an existing system.\n",
    "Throughout the code, we have left comments where phylogeny tracking code should be added.\n",
    "Below, we link to some existing examples and the `phylotrackpy` documentation to \n",
    "for your reference to get tracking working. \n",
    "If you're working on this during the tutorial at ALife 2024, feel free to ask for our help! \n",
    "We're happy to walk you through anything!  \n",
    "\n",
    "## Helpful reference material\n",
    "\n",
    "- `phylotrackpy` documentation: <https://phylotrackpy.readthedocs.io/en/latest/introduction.html>\n",
    "- `phylotrackpy` GitHub repository: <https://github.com/emilydolson/phylotrackpy>\n",
    "- Example code from an ALife 2023 tutorial: <https://github.com/emilydolson/alife-phylogeny-tutorial/blob/main/perfect_tracking_final.ipynb> \n",
    "  - Scroll down to the `phylotrackpy` heading!\n",
    "- A completed version of this notebook with phylogeny tracking implemented already: <https://github.com/amlalejini/alife-2024-phylo-tutorial/blob/main/notebooks/phylotrackpy-sync-gen-example-completed.ipynb>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup\n",
    "\n",
    "First, install required Python packages for this example (e.g., phylotrackpy).\n",
    "\n",
    "### Local setup\n",
    "\n",
    "If you are running this locally, we recommend making a Python virtual environment to keep your local Python installation clean: \n",
    "\n",
    "```\n",
    "python -m venv venv\n",
    "source venv/bin/activate\n",
    "!python -m pip install -r requirements.txt\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Google Colab setup\n",
    "\n",
    "If you're running this on Google Colab, you can run the following python cell to install the requisite packages!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!python -m pip install phylotrackpy\n",
    "!python -m pip install polars       # Used by example for data tracking"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "from phylotrackpy import systematics\n",
    "import random\n",
    "import polars as pl\n",
    "\n",
    "# Seed random number generator\n",
    "random.seed(8)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## System implementation\n",
    "\n",
    "Here, the `Organism` class (below) defines candidate solutions with binary genomes, an evaluated fitness value, and a taxon ID. \n",
    "The taxon ID member variable keeps track of the organism's taxon in the phylogeny, which is helpful for `phylotrackpy`'s phylogeny tracker.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Organism:\n",
    "    def __init__(\n",
    "        self,\n",
    "        num_genes:int = 10,\n",
    "        randomize_genome:bool = False\n",
    "    ):\n",
    "        # Genomes are vectors of binary values\n",
    "        self.genome = [bool(random.randint(0, 1)) if randomize_genome else False for _ in range(num_genes)]\n",
    "        # Evaluate initial fitness\n",
    "        self.EvalFitness()\n",
    "        # Organisms keep track of their taxon id (useful for phylogeny tracking)\n",
    "        self.taxon_id = None\n",
    "\n",
    "    @classmethod\n",
    "    def FromGenome(cls, genome:list):\n",
    "        # Create new organism from a given genome.\n",
    "        org = cls(\n",
    "            num_genes = 0,\n",
    "            randomize_genome = False\n",
    "        )\n",
    "        org.SetGenome(genome)\n",
    "        org.EvalFitness()\n",
    "        return org\n",
    "\n",
    "    def GetGenome(self):\n",
    "        return self.genome\n",
    "\n",
    "    def SetGenome(self, genome:list):\n",
    "        self.genome = [gene for gene in genome]\n",
    "        # Evaluate fitness after updating genome\n",
    "        self.EvalFitness()\n",
    "\n",
    "    def GetFitness(self):\n",
    "        return self.fitness\n",
    "\n",
    "    def EvalFitness(self):\n",
    "        self.fitness = sum(self.genome)\n",
    "        return self.fitness\n",
    "\n",
    "    def GetTaxonID(self):\n",
    "        return self.taxon_id\n",
    "\n",
    "    def SetTaxonID(self, id):\n",
    "        self.taxon_id = id\n",
    "\n",
    "    def Mutate(self, per_site_mut_rate:float=0.01):\n",
    "        num_muts = 0\n",
    "        for gene_i in range(len(self.genome)):\n",
    "            if (random.random() <= per_site_mut_rate):\n",
    "                self.genome[gene_i] = not self.genome[gene_i]\n",
    "                num_muts += 1\n",
    "        # Update fitness evaluation after mutations\n",
    "        self.EvalFitness()\n",
    "        return num_muts\n",
    "\n",
    "\n",
    "def TournamentSelect(num_parents:int, population:list, tourn_size:int = 4):\n",
    "    parent_ids = [None for _ in range(num_parents)]\n",
    "    candidate_ids = [i for i in range(len(population))]\n",
    "    for i in range(num_parents):\n",
    "        tourn_participants = random.sample(candidate_ids, tourn_size)\n",
    "        parent_ids[i] = max([(population[idx].GetFitness(), idx) for idx in tourn_participants])[1]\n",
    "    return parent_ids\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here, we implement a basic evolutionary computing loop solving the one-max problem to demonstrate how `phylotrackpy` can be integrated into a traditional evolutionary computing context.\n",
    "\n",
    "**We've marked locations where phylogeny tracking code needs to be added with \"TUTORIAL TASK\" comments!**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pop_size = 500\n",
    "generations = 1000\n",
    "per_site_mut_rate = 0.05\n",
    "genome_length = 100\n",
    "tournament_size = 8\n",
    "print_resolution = 100\n",
    "summary_output_resolution = 10\n",
    "\n",
    "assert(pop_size > 0)\n",
    "assert(generations > 0)\n",
    "assert(per_site_mut_rate >= 0 and per_site_mut_rate <= 1.0)\n",
    "assert(genome_length > 0)\n",
    "assert(print_resolution > 0)\n",
    "assert(summary_output_resolution > 0)\n",
    "\n",
    "'''\n",
    "TUTORIAL TASK: Initialization\n",
    "- (1) Create a new systematics object here to track the population's phylogeny\n",
    "- (2) Add snapshot functions that capture basic information about a taxon (ask for clarification if you're not sure what this means)\n",
    "- (3) Initialize the systematics object's update to 0\n",
    "'''\n",
    "\n",
    "# Create a list to hold phylo metrics over time\n",
    "data = []\n",
    "\n",
    "# Create initial population\n",
    "population = [Organism(num_genes=genome_length, randomize_genome=False) for _ in range(pop_size)]\n",
    "\n",
    "'''\n",
    "TUTORIAL TASK: Add initial population to the phylogeny\n",
    "- After adding the initial individuals to the phylogeny, store each taxon id on the individual\n",
    "'''\n",
    "\n",
    "for gen in range(0, generations+1):\n",
    "    if (gen % print_resolution) == 0:\n",
    "        print(f\"---Update {gen}---\")\n",
    "        print(f\"  Max fitness={max([indiv.GetFitness() for indiv in population])}\")\n",
    "\n",
    "    '''\n",
    "    TUTORIAL TASK: Update the systematics object's update to current generation\n",
    "    '''\n",
    "\n",
    "    # Individuals are evaluated immediately after reproduction, so no need to evaluate\n",
    "    # the full population here. In other systems, population evaulation would likely happen here.\n",
    "\n",
    "    # Select parents to reproduce.\n",
    "    parent_ids = TournamentSelect(\n",
    "        num_parents = pop_size,\n",
    "        population = population,\n",
    "        tourn_size = tournament_size\n",
    "    )\n",
    "\n",
    "    # Create next generation's population by copying parents, mutating offspring,\n",
    "    # and adding offspring to the phylogeny.\n",
    "    next_pop = []\n",
    "    for parent_id in parent_ids:\n",
    "        next_pop.append(Organism.FromGenome(population[parent_id].GetGenome()))\n",
    "        next_pop[-1].Mutate(per_site_mut_rate)\n",
    "        '''\n",
    "        TUTORIAL TASK: Add new offspring to phylogeny\n",
    "        - (1) Add offspring to phylogeny\n",
    "        - (2) Update offspring's taxon id\n",
    "        '''\n",
    "\n",
    "    # Record data for this update\n",
    "    if (gen % summary_output_resolution) == 0:\n",
    "        '''\n",
    "        TUTORIAL TASK: Add some phylogeny metrics to the data being saved\n",
    "        '''\n",
    "        data.append({\n",
    "            \"generation\": gen,\n",
    "            \"max_fit\": max([indiv.GetFitness() for indiv in population])\n",
    "        })\n",
    "\n",
    "    '''\n",
    "    TUTORIAL TASK: Remove current population individuals from the phylogeny before we replace them with next_pop\n",
    "    '''\n",
    "    population = next_pop\n",
    "\n",
    "'''\n",
    "TUTORIAL TASK: Output a final phylogeny snapshot\n",
    "'''\n",
    "summary_df = pl.DataFrame(data)\n",
    "summary_df.write_csv(\"summary.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Visualization\n",
    "\n",
    "Phylogeny visualization is an area of active research, particularly as it applies to the more complex phylogenies that we often get in ALife research. Here are some suggestions.\n",
    "\n",
    "## ALife Phylogeny Visualizer\n",
    "\n",
    "https://emilydolson.github.io/lineage_viz_tool/phylogeny_visualizations/phylogeny.html\n",
    "\n",
    "This tool is built for visualizing artificial life phylogenies. It is written in Javascript. It expects data in [ALife standard phylogeny data format](https://alife-data-standards.github.io/alife-data-standards/phylogeny). You can upload a file using the file selector, and it should immediately visualize your phylogeny. From there, there are a few settings that you can adjust:\n",
    "\n",
    "- Scale exponent: Under some evolutionary regimes, most of the interesting phylogenetic structure happens very near the end of evolutionary times. In other cases, it happens early on. To accommodate these differences, the visualizer uses an exponentially-scaled time axis. To adjust the exponent used (and therefore the temporal part of the phylogeny being emphasized), slide the slider.\n",
    "\n",
    "- Show intermediate taxa: In biology, usually only the tips (i.e. leaf nodes) of the phylogeny are known. In ALife, we usually know details about every node of the phylogeny. We can take advantage of this information. By toggling on this checkbox, the visualizer will display rectangles alongside the phylogeny depicting the portion of evolutionary time that each taxon was alive for. Sometimes, the lineage that lead to an extant taxon will contain many taxa that coexisted with each other. In these cases, rectangles are drawn nexxt to each other.\n",
    "\n",
    "- You can also adjust the colors of the lines and circles in the phylogeny.\n",
    "\n",
    "## IcyTree\n",
    "\n",
    "https://icytree.org/\n",
    "\n",
    "This is a visualizer built for bioinformaticians. Consequently, it lacks the ALife-specific features. However, it is more mature and contains lots of useful features. To use it, you can use the alife-phyloinformatics-convert python package to convert an ALife Standard Phylogeny file to a bioinformatics standard such as newick. Note that newick format does not always play nicely with very large phylogenies, so you may need to sub-sample yours to get a good visualization.\n",
    "\n",
    "The following code will convert an ALife standard phylogeny to newick"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import alifedata_phyloinformatics_convert as apc\n",
    "\n",
    "# Optional: if your data isn't already in a phylotrackpy systematics \n",
    "# object, load it in\n",
    "# (only do this with an empty systematics pbject)\n",
    "#\n",
    "# phylo_tracker.load_from_file(\"phylo_snapshot_final.csv\")\n",
    "\n",
    "\n",
    "converter = apc.RosettaTree(phylo_tracker)\n",
    "newick_string = converter.to_newick()  # returns newick string\n",
    "with open(\"phylogeny.nwk\", \"w\") as f:\n",
    "    f.write(newick_string)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can then upload this newick file to IcyTree!"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pyenv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
