{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a target=\"_blank\" href=\"https://colab.research.google.com/github/amlalejini/alife-2024-phylo-tutorial/blob/main/notebooks/phylotrackpy-async-gen-example-completed.ipynb\">\n",
    "  <img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/>\n",
    "</a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Phylotrackpy with asynchronous generations example (completed)\n",
    "\n",
    "This notebook contains a very simple evolving system with asynchronous generations \n",
    "wherein organisms reproduce as they accumulate the requisite resources.\n",
    "Each organism's genome is a binary string, and organism's accumulate resources as \n",
    "a function of the proportion of 1s in their genome. \n",
    "I.e., this is a one-max environment!  \n",
    "\n",
    "The example code in this notebook incorporates phylogeny tracking.\n",
    "Below, we link to some existing examples and the `phylotrackpy` documentation to \n",
    "for your reference. \n",
    "If you're working on this during the tutorial at ALife 2024, feel free to ask for our help in understanding the code! \n",
    "We're happy to walk you through anything!  \n",
    "\n",
    "## Helpful reference material\n",
    "\n",
    "- `phylotrackpy` documentation: <https://phylotrackpy.readthedocs.io/en/latest/introduction.html>\n",
    "- `phylotrackpy` GitHub repository: <https://github.com/emilydolson/phylotrackpy>\n",
    "- Example code from an ALife 2023 tutorial: <https://github.com/emilydolson/alife-phylogeny-tutorial/blob/main/perfect_tracking_final.ipynb> \n",
    "  - Scroll down to the `phylotrackpy` heading!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup\n",
    "\n",
    "First, install required Python packages for this example (e.g., phylotrackpy).\n",
    "\n",
    "### Local setup\n",
    "\n",
    "If you are running this locally, we recommend making a Python virtual environment to keep your local Python installation clean: \n",
    "\n",
    "```\n",
    "python -m venv venv\n",
    "source venv/bin/activate\n",
    "!python -m pip install -r requirements.txt\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Google Colab setup\n",
    "\n",
    "If you're running this on Google Colab, you can run the following python cell to install the requisite packages!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!python -m pip install phylotrackpy\n",
    "!python -m pip install polars       # Used by example for data tracking"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from phylotrackpy import systematics\n",
    "import random\n",
    "import polars as pl\n",
    "\n",
    "# Seed random number generator\n",
    "random.seed(8)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## System implementation\n",
    "\n",
    "The `Organism` class (below) defines organisms with a binary genome, resource count, and taxon id.\n",
    "The taxon id member keeps track of the organism's taxon in the phylogeny (which is helpful for `phylotrackpy`'s phylogeny tracker)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Organism:\n",
    "    def __init__(\n",
    "        self,\n",
    "        num_genes:int = 10,\n",
    "        randomize_genome:bool = False,\n",
    "        init_resources:float = 0.0\n",
    "    ):\n",
    "        # Genomes are vectors of binary values\n",
    "        self.genome = [bool(random.randint(0, 1)) if randomize_genome else False for _ in range(num_genes)]\n",
    "        # Organisms have resources that determine whether they can reproduce\n",
    "        self.resources = init_resources\n",
    "        # Organisms keep track of their taxon id (useful for phylogeny tracking)\n",
    "        self.taxon_id = None\n",
    "\n",
    "    @classmethod\n",
    "    def FromGenome(cls, genome:list):\n",
    "        # Create new organism from a given genome.\n",
    "        org = cls(\n",
    "            num_genes = 0,\n",
    "            randomize_genome = False,\n",
    "            init_resources = 0.0\n",
    "        )\n",
    "        org.SetGenome(genome)\n",
    "        return org\n",
    "\n",
    "    def GetGenome(self):\n",
    "        return self.genome\n",
    "\n",
    "    def SetGenome(self, genome:list):\n",
    "        self.genome = [gene for gene in genome]\n",
    "\n",
    "    def GetResources(self):\n",
    "        return self.resources\n",
    "\n",
    "    def SetResources(self, amount:float):\n",
    "        self.resources = amount\n",
    "\n",
    "    def IncResources(self, amount:float=1.0):\n",
    "        self.resources += amount\n",
    "\n",
    "    def DecResources(self, amount:float=1.0):\n",
    "        self.resources -= amount\n",
    "\n",
    "    def GetTaxonID(self):\n",
    "        return self.taxon_id\n",
    "\n",
    "    def SetTaxonID(self, id):\n",
    "        self.taxon_id = id\n",
    "\n",
    "    def Mutate(self, per_site_mut_rate:float=0.01):\n",
    "        num_muts = 0\n",
    "        for gene_i in range(len(self.genome)):\n",
    "            if (random.random() <= per_site_mut_rate):\n",
    "                self.genome[gene_i] = not self.genome[gene_i]\n",
    "                num_muts += 1\n",
    "        return num_muts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The \"world\" implementation is below. Organisms exist in a \"well-mixed\" environment (the `world` list).\n",
    "On each update, an organism is chosen at random to \"execute\", gaining resources proportional to the number of 1s in its genome. \n",
    "That organism then reproduces if it has sufficient resources. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# World parameters\n",
    "max_world_size = 500           # Determines maximum number of organisms in world\n",
    "updates = 20000                 # How long to run system?\n",
    "repro_cost = 10.0              # Cost of reproduction\n",
    "max_resource_intake = 5        # Maximum amount of resources an organism can gain when \"executed\"\n",
    "per_site_mut_rate = 0.05       # Per-site mutation rate for offspring\n",
    "genome_length = 100            # Length of organism genomes\n",
    "print_resolution = 100         # Print world status at this interval\n",
    "snapshot_resolution = 2000     # Snapshot the phylogeny at this interval\n",
    "summary_output_resolution = 10 # Interval to save summary information\n",
    "\n",
    "assert(max_world_size > 0)\n",
    "assert(updates > 0)\n",
    "assert(per_site_mut_rate >= 0 and per_site_mut_rate <= 1.0)\n",
    "assert(genome_length > 0)\n",
    "\n",
    "# Create a new Systematics object to track the population's phylogeny\n",
    "# - Taxa in the phylogeny will be defined by genotypes (i.e., offspring with the same genotype as their parent will belong to the same taxon).\n",
    "phylo_tracker = systematics.Systematics(\n",
    "    lambda org: org.GetGenome(),\n",
    "    store_pos = False\n",
    ")\n",
    "# Configure snapshots to output basic information about a taxon\n",
    "phylo_tracker.add_snapshot_fun(\n",
    "    lambda tax: str(list(map(int,tax.get_info()))).replace(\",\", \" \"),\n",
    "    \"genome\",\n",
    "    \"Taxon's genome\"\n",
    ")\n",
    "phylo_tracker.add_snapshot_fun(\n",
    "    lambda tax: str(sum(tax.get_info())),\n",
    "    \"fitness\",\n",
    "    \"Number of ones in taxon genome\"\n",
    ")\n",
    "# Initialize phylogeny tracker's update to 0\n",
    "phylo_tracker.set_update(0)\n",
    "\n",
    "# Create a list to hold phylo metrics over time\n",
    "data = []\n",
    "\n",
    "# Create population seeded with initial organism\n",
    "world = [Organism(num_genes=genome_length, randomize_genome=False)]\n",
    "# Add initial organism to phylogeny tracker\n",
    "world[0].taxon_id = phylo_tracker.add_org(world[0])\n",
    "\n",
    "# Execute world for configured number of updates\n",
    "for update in range(0, updates+1):\n",
    "    if (update % print_resolution) == 0:\n",
    "        print(f\"---Update {update}---\")\n",
    "        print(f\"  Population size={len(world)}\")\n",
    "        print(f\"  Max fitness={max([sum(world[i].GetGenome()) for i in range(len(world))])}\")\n",
    "\n",
    "    # Set phylo tracker's update to current world update\n",
    "    phylo_tracker.set_update(update)\n",
    "\n",
    "    # Get ID of individual to \"execute\"\n",
    "    pop_id = random.randrange(0, len(world)) if len(world) > 1 else 0\n",
    "    cur_org = world[pop_id]\n",
    "\n",
    "    # Calculate resource gain\n",
    "    fit = sum(cur_org.GetGenome())\n",
    "    resource_gain = max((fit / genome_length) * max_resource_intake, 0.1)\n",
    "    cur_org.IncResources(resource_gain)\n",
    "\n",
    "    # Check if organism has resources to reproduce\n",
    "    if cur_org.GetResources() >= repro_cost:\n",
    "        # Current organism pays cost of reproduction\n",
    "        cur_org.DecResources(repro_cost)\n",
    "        # Create offspring as copy of current organism\n",
    "        offspring = Organism.FromGenome(cur_org.GetGenome())\n",
    "        mut_count = offspring.Mutate(per_site_mut_rate)\n",
    "        # Add offspring to phylogeny tracker, note its taxon ID\n",
    "        offspring_taxon_id = phylo_tracker.add_org(offspring, cur_org.taxon_id)\n",
    "        offspring.SetTaxonID(offspring_taxon_id)\n",
    "        # Place offspring into world\n",
    "        if len(world) < max_world_size:\n",
    "            # If world not full, just append.\n",
    "            world.append(offspring)\n",
    "        else:\n",
    "            # Otherwise, random replacement.\n",
    "            offspring_loc = random.randint(0, len(world)-1)\n",
    "            # Mark organism to be removed from phylogeny for deletion\n",
    "            phylo_tracker.remove_org(world[offspring_loc].GetTaxonID())\n",
    "            # Offspring replaces organism at chosen location.\n",
    "            world[offspring_loc] = offspring\n",
    "\n",
    "    # Snapshot phylogeny\n",
    "    if (update % snapshot_resolution) == 0:\n",
    "        phylo_tracker.snapshot(f\"phylo_snapshot_{update}.csv\")\n",
    "\n",
    "    # Record data for this update\n",
    "    if (update % summary_output_resolution) == 0:\n",
    "        data.append({\n",
    "            \"update\": update,\n",
    "            \"pop_size\": len(world),\n",
    "            \"phylo_num_taxa\": phylo_tracker.get_num_taxa(),\n",
    "            \"phylo_sum_pairwise_dist\": phylo_tracker.get_sum_pairwise_distance(False),\n",
    "            \"phylo_variance_pairwise_dist\": phylo_tracker.get_variance_pairwise_distance(False),\n",
    "            \"phylo_phylogenetic_diversity\": phylo_tracker.get_phylogenetic_diversity(),\n",
    "            \"phylo_mrca_depth\": phylo_tracker.mrca_depth()\n",
    "        })\n",
    "\n",
    "# Optional: These phylogenies are a little big to visualize! Visualizing large\n",
    "# phylogenies is an area of active research. For now, a good strategy is to\n",
    "# sample the leaf nodes. The code below does that\n",
    "leaves_to_retain = 10\n",
    "taxa_to_remove = random.sample(phylo_tracker.get_active_taxa(), phylo_tracker.get_num_active() - leaves_to_retain)\n",
    "for t in taxa_to_remove:\n",
    "    for i in range(t.get_num_orgs()):\n",
    "        phylo_tracker.remove_org(t)\n",
    "\n",
    "\n",
    "phylo_tracker.snapshot(\"phylo_snapshot_final.csv\")\n",
    "summary_df = pl.DataFrame(data)\n",
    "summary_df.write_csv(\"phylo_summary.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Visualization\n",
    "\n",
    "Phylogeny visualization is an area of active research, particularly as it applies to the more complex phylogenies that we often get in ALife research. Here are some suggestions.\n",
    "\n",
    "## ALife Phylogeny Visualizer\n",
    "\n",
    "https://emilydolson.github.io/lineage_viz_tool/phylogeny_visualizations/phylogeny.html\n",
    "\n",
    "This tool is built for visualizing artificial life phylogenies. It is written in Javascript. It expects data in [ALife standard phylogeny data format](https://alife-data-standards.github.io/alife-data-standards/phylogeny). You can upload a file using the file selector, and it should immediately visualize your phylogeny. From there, there are a few settings that you can adjust:\n",
    "\n",
    "- Scale exponent: Under some evolutionary regimes, most of the interesting phylogenetic structure happens very near the end of evolutionary times. In other cases, it happens early on. To accommodate these differences, the visualizer uses an exponentially-scaled time axis. To adjust the exponent used (and therefore the temporal part of the phylogeny being emphasized), slide the slider.\n",
    "\n",
    "- Show intermediate taxa: In biology, usually only the tips (i.e. leaf nodes) of the phylogeny are known. In ALife, we usually know details about every node of the phylogeny. We can take advantage of this information. By toggling on this checkbox, the visualizer will display rectangles alongside the phylogeny depicting the portion of evolutionary time that each taxon was alive for. Sometimes, the lineage that lead to an extant taxon will contain many taxa that coexisted with each other. In these cases, rectangles are drawn nexxt to each other.\n",
    "\n",
    "- You can also adjust the colors of the lines and circles in the phylogeny.\n",
    "\n",
    "## IcyTree\n",
    "\n",
    "https://icytree.org/\n",
    "\n",
    "This is a visualizer built for bioinformaticians. Consequently, it lacks the ALife-specific features. However, it is more mature and contains lots of useful features. To use it, you can use the alife-phyloinformatics-convert python package to convert an ALife Standard Phylogeny file to a bioinformatics standard such as newick. Note that newick format does not always play nicely with very large phylogenies, so you may need to sub-sample yours to get a good visualization.\n",
    "\n",
    "The following code will convert an ALife standard phylogeny to newick"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!python -m pip install alifedata_phyloinformatics_convert"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "import alifedata_phyloinformatics_convert as apc\n",
    "\n",
    "# Optional: if your data isn't already in a phylotrackpy systematics \n",
    "# object, load it in\n",
    "# (only do this with an empty systematics pbject)\n",
    "#\n",
    "# phylo_tracker.load_from_file(\"phylo_snapshot_final.csv\")\n",
    "\n",
    "\n",
    "converter = apc.RosettaTree(phylo_tracker)\n",
    "newick_string = converter.to_newick()  # returns newick string\n",
    "with open(\"phylogeny.nwk\", \"w\") as f:\n",
    "    f.write(newick_string)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can then upload this newick file to IcyTree!"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
